### PR TITLE
Generated clients can return their RESTClients, RESTClient can return its RateLimiter

### PR DESCRIPTION
**Pull request by gmarek**
cc @lavalamp @krousey @wojtek-t @smarterclayton @timothysc 

Ref. #22421

### CODE REVIEW

**Comment by timothysc** [comment]
@gmarek I'm sorry, why are throttling if no value is set from the client?  I'm missing the motivation/problem statement on this...

**Comment by gmarek** [comment]
@timothysc - which thing you mean? Is this a first or a second commit? First actually has a separate PR #24166.

**Comment by gmarek** [comment]
@lavalamp PTAL, @timothysc can you explain what do you mean?

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 1]
```diff
@@ -68,7 +69,9 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
 		token:  options.TokenGenerator,
 		rootCA: options.RootCA,
 	}
-
+	if cl != nil && cl.Core() != nil && cl.Core().GetRESTClient() != nil {
+		metrics.RegisterMetricAndTrackRateLimitierUsage("serviceaccount_controller", cl.Core().GetRESTClient().GetRateLimiter())
```
Comment (lavalamp) [inline pkg/controller/serviceaccount/tokens_controller.go]: Somewhere, you need a test that the client.GetRESTClient().GetRateLimiter() call sequence actually gets the right rate limiter. I can see that getting broken in the future (by e.g. someone modifying a client's RESTClient or something)

**Comment by lavalamp** [comment]
I didn't do a deep pass. I'll let @krousey do the final review.

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 1 above)
Comment (gmarek) [inline pkg/controller/serviceaccount/tokens_controller.go]: Good point. I'll add that.

**Comment by gmarek** [comment]
@lavalamp - added a test to generated_client directory - I had no better idea where to place it. If you think there's a better place for it let me know. I'm also not sure if this requires release note.

@krousey PTAL

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go [hunk 2]
```diff
@@ -51,16 +51,19 @@ func (g *genFakeForGroup) Namers(c *generator.Context) namer.NameSystems {
 }
 
 func (g *genFakeForGroup) Imports(c *generator.Context) (imports []string) {
-	return append(g.imports.ImportLines(), fmt.Sprintf("%s \"%s\"", filepath.Base(g.realClientPath), g.realClientPath))
+	imports = append(g.imports.ImportLines(), fmt.Sprintf("%s \"%s\"", filepath.Base(g.realClientPath), g.realClientPath))
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go]: Just out of curiosity, why did you change this?

File: pkg/client/restclient/client.go [hunk 3]
```diff
@@ -111,6 +111,10 @@ func readExpBackoffConfig() BackoffManager {
 			time.Duration(backoffDurationInt)*time.Second)}
 }
 
+func (c *RESTClient) GetRateLimiter() flowcontrol.RateLimiter {
```
Comment (krousey) [inline pkg/client/restclient/client.go]: I don't understand why this is needed since `Throttle` is already public.

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go (hunk 2 above)
Comment (gmarek) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go]: I though I need to add an import, but it turned out that I don't. I can revert it if you think it's worth it.

File: pkg/client/restclient/client.go (hunk 3 above)
Comment (gmarek) [inline pkg/client/restclient/client.go]: Right - it's not, originally I was doing something in this function, but in current state I can just remove it.

File: pkg/controller/daemon/controller.go [hunk 4]
```diff
@@ -113,6 +114,9 @@ func NewDaemonSetsController(podInformer framework.SharedInformer, kubeClient cl
 	// TODO: remove the wrapper when every clients have moved to use the clientset.
 	eventBroadcaster.StartRecordingToSink(&unversionedcore.EventSinkImpl{Interface: kubeClient.Core().Events("")})
 
+	if kubeClient != nil && kubeClient.Core() != nil && kubeClient.Core().GetRESTClient() != nil {
```
Comment (krousey) [inline pkg/controller/daemon/controller.go]: This would look a lot cleaner if `Core()`, `GetRESTClient()` and `GetRateLimiter()` handled a `nil` receiver and returned `nil`. Then you could just write:

```go
if rl := kubeClient.Core().GetRESTClient().GetRateLimiter(); rl != nil {
	metrics.RegisterMetricAndTrackRateLimiterUsage("daemon_controller", rl)
}
```

File: pkg/util/metrics/util.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		Help:      fmt.Sprintf("A metric measuring the saturation of the rate limiter for %v", ownerName),
+	})
+	rateLimiterMetrics[ownerName] = metric
+	prometheus.MustRegister(metric)
+	return nil
+}
+
+// RegisterMetricAndTrackRateLimitierUsage registers a metric ownerName_rate_limiter_use in prometheus to track
+// how much used rateLimiter is and starts a goroutine that updates this metric every updatePeriod
+func RegisterMetricAndTrackRateLimitierUsage(ownerName string, rateLimiter flowcontrol.RateLimiter) error {
```
Comment (krousey) [inline pkg/util/metrics/util.go]: Misspelled limiter

Comment (gmarek) [inline pkg/util/metrics/util.go]: Done.

File: pkg/controller/daemon/controller.go (hunk 4 above)
Comment (gmarek) [inline pkg/controller/daemon/controller.go]: Done as much as possible. Doesn't work for Core (as we use clientset as an interface, and I'm not sure if it's possible to have a nil-implementations of interfaces).

**Comment by gmarek** [comment]
@krousey PTAL

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go (hunk 2 above)
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_group.go]: meh.. It's just noise in the review to me. I had to think about it far longer than I would have liked because I didn't think it really changed anything.

File: cmd/libs/go2idl/client-gen/generators/generator_for_group.go [hunk 6]
```diff
@@ -168,6 +170,17 @@ func NewForConfigOrDie(c *$.Config|raw$) *$.Group$Client {
 }
 `
 
+var getRESTClient = `
+// GetRESTClient returns a RESTClient that is used to communicate
+// with API server by this client implementation.
+func (c *$.Group$Client) GetRESTClient() *$.RESTClient|raw$ {
+  if c == nil {
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/generator_for_group.go]: This indentation is wrong.

File: pkg/util/metrics/util.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		Name:      "rate_limiter_use",
+		Subsystem: ownerName,
+		Help:      fmt.Sprintf("A metric measuring the saturation of the rate limiter for %v", ownerName),
+	})
+	rateLimiterMetrics[ownerName] = metric
+	prometheus.MustRegister(metric)
+	return nil
+}
+
+// RegisterMetricAndTrackRateLimitierUsage registers a metric ownerName_rate_limiter_use in prometheus to track
```
Comment (krousey) [inline pkg/util/metrics/util.go]: It's misspelled here too.

**Comment by krousey** [comment]
2 minor nits, but LGTM after they're fixed.

File: cmd/libs/go2idl/client-gen/generators/generator_for_group.go (hunk 6 above)
Comment (gmarek) [inline cmd/libs/go2idl/client-gen/generators/generator_for_group.go]: Why aren't we using spaces here as we do everywhere else...

**Comment by gmarek** [comment]
Fixed. @krousey PTAL.

**Comment by gmarek** [comment]
I need to fix the test, after I moved it to a different directory:/

**Comment by gmarek** [comment]
@krousey PTAL

**Comment by smarterclayton** [comment]
I may suggest a refactor for 1.4, where we provide this as a config wrapper for the rest client.  We have to jump through some hoops to expose this out in a clean way.  We may want a client factory for use by infra components, for example.
